lldb/source/Plugins/LanguageRuntime/Swift/ReflectionContextInterface.h

```diff
@@ -0,0 +1,147 @@
+    virtual const swift::reflection::TypeRef *
+    GetTypeRefOrNull(swift::Demangle::Demangler &dem,
+                     swift::Demangle::NodePointer node) = 0;
+    virtual const swift::reflection::TypeInfo *
+    GetClassInstanceTypeInfo(const swift::reflection::TypeRef *type_ref,
+                             swift::remote::TypeInfoProvider *provider) = 0;
+    virtual const swift::reflection::TypeInfo *
+    GetTypeInfo(const swift::reflection::TypeRef *type_ref,
+                swift::remote::TypeInfoProvider *provider) = 0;
+    virtual const swift::reflection::TypeInfo *
+    GetTypeInfoFromInstance(lldb::addr_t instance,
+                            swift::remote::TypeInfoProvider *provider) = 0;
+    virtual swift::remote::MemoryReader &GetReader() = 0;
+    virtual const swift::reflection::TypeRef *
+    LookupSuperclass(const swift::reflection::TypeRef *tr) = 0;
+    virtual bool
+    ForEachSuperClassType(swift::remote::TypeInfoProvider *tip,
+                          lldb::addr_t pointer,
+                          std::function<bool(SuperClassType)> fn) = 0;
+    virtual llvm::Optional<std::pair<const swift::reflection::TypeRef *,
+                                     swift::remote::RemoteAddress>>
+    ProjectExistentialAndUnwrapClass(
+        swift::remote::RemoteAddress existential_addess,
+        const swift::reflection::TypeRef &existential_tr) = 0;
+    virtual llvm::Optional<int32_t>
+    ProjectEnumValue(swift::remote::RemoteAddress enum_addr,
+                     const swift::reflection::TypeRef *enum_type_ref,
+                     swift::remote::TypeInfoProvider *provider) = 0;
+    virtual const swift::reflection::TypeRef *
+    ReadTypeFromMetadata(lldb::addr_t metadata_address,
+                         bool skip_artificial_subclasses = false) = 0;
+    virtual const swift::reflection::TypeRef *
+    ReadTypeFromInstance(lldb::addr_t instance_address,
+                         bool skip_artificial_subclasses = false) = 0;
+    virtual llvm::Optional<bool> IsValueInlinedInExistentialContainer(
+        swift::remote::RemoteAddress existential_address) = 0;
+    virtual const swift::reflection::TypeRef *
+    ApplySubstitutions(const swift::reflection::TypeRef *type_ref,
+                       swift::reflection::GenericArgumentMap substitutions) = 0;
+    virtual swift::remote::RemoteAbsolutePointer
+    StripSignedPointer(swift::remote::RemoteAbsolutePointer pointer) = 0;
+    virtual swift::reflection::TypeRefBuilder &GetBuilder() = 0;
+  };
+
+
+  /// A wrapper around TargetReflectionContext, which holds a lock to ensure
+  /// exclusive access.
+  struct ThreadSafeReflectionContext {
+    ThreadSafeReflectionContext(ReflectionContextInterface *reflection_ctx,
+                             std::recursive_mutex &mutex)
+        : m_reflection_ctx(reflection_ctx), m_lock(mutex, std::adopt_lock) {}
+
+    static ThreadSafeReflectionContext MakeInvalid() {
```

Comment:
@adrian-prantl @kastiglione I'm not really happy with this, but it's required for returning the "empty" case in the stub language runtime. If you have any ideas on how to make this better I'd be happy to listen.